Create CBMC developer guide documentation

owen-mc-diffblue opened the pull request:
Currently this is an outline - over time the various sections should
be filled in. I have moved some of the existing documentation into it
in the appropriate place.

Commit:
Create CBMC developer guide documentation

Currently this is an outline - over time the various sections should
be filled in. I have moved some of the existing documentation into it
in the appropriate place.

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+
+  subgraph top {
+    rank=same;
+    1 -> 2 -> 3 -> 4;
```
You can give these informative short names instead of 1, 2, 3, ...

owen-mc-diffblue replied:
I didn't write this. It seems clear enough to me.

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+transformation tools (see \ref section-other-tools).
+
+# Concepts #
+## {C, java bytecode} -> Parse tree -> Symbol table -> GOTO programs → GOTO program transformations → BMC → counterexample (goto_tracet) → printing ##
```
Some ascii arrows, some unicode ones

owen-jones-diffblue replied:
It doesn't seem to affect the output. I'll standardise on ascii ones.

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+
+FIXME
+
+## Instrumentation (for test gen & CBMC & …): Goto functions -> goto functions ##
```
Similarly pick between ASCII and unicode ellipses

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -70,12 +255,12 @@ In the top level of `src` there are only a few files:
 * `doxygen.cfg`:   The config file for doxygen.cfg
 
-## `doc/`
+## `doc/` ##
```
FWIW I think this is unnecessary in markdown

owen-mc-diffblue replied:
Yeah, I discovered it yesterday and thought it might make the titles stand out more when looking at the source

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
-some cases the algorithms used are described in the relevant papers.
+## Strings: dstringt, the string_container and the ID_* ##
+Within cbmc, strings are represented using irep_idt. By default this is
+typedefed to dstringt, which stores a string as an index into a large
```
Probably use backticks for typenames etc

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
+this irep_idt as ID_type. The other kind of line you see is
+“IREP_ID_TWO(C_source_location, #source_location)”, which means the
+irep_idt for the string “#source_location” can be referred to as
+ID_C_source_location. The “C” is for comment, which will be explained
```
Worth mentioning that while it is init'd from `irep_ids.def` it also grows dynamically, so `irep_idt i1("Hello"), i2("Hello")` does produce the same table index twice. Compare `irept ir1("hello"), ir2("hello")`, where the ireps will share the `irep_idt` but the `irept` itself will not be shared.

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
-CBMC takes C code or a goto-binary as input and tries to emit traces
-of executions that lead to crashes or undefined behaviour. The diagram
-below shows the intermediate steps in this process.
+## Irept: a 4-triple (data, named-sub, comments-sub, sub) ##
```
Suggest put `sub` after `data` as the most-used part

owen-mc-diffblue replied:
We should probably be consistent with the order they are listed in the documentation for `irept` (cf #2618), which is this order. It's also the order that they're listed in in `irep.h`.

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
 
-\dot
-digraph G {
+As that documentation says, irepts are generic tree nodes. You should
```
Mention that `data` is just a string (`irep_idt`)

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
+so a minus_exprt has ID_minus as its id(). This means that a
+minus_exprt can be passed wherever an exprt is expected, and if you
+want to check if the expression you are looking at is a minus
+expression then you have to check its id().
```
...or use `can_cast_expr` and cousins

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
+Many different kinds of statements inherit from codet, and they are
+distinguished by their statement(). For example, a while loop would be
+represented by a code_whilet, which has statement() ID_while.
+Code_whilet has two operands in its sub, and helper functions to make
```
Don't capitalise types, even at start of sentence

smowton commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
+Code_whilet has two operands in its sub, and helper functions to make
+it easier to use: cond() returns the first entry in its sub(), which
+is the condition for the while loop, and body() returns the second
+entry in its sub, which is the body of the while loop - this has to be
```
Worth defining and then using the terminology "first subexpression, second subexpression, ..." rather than "... in its sub"?

peterschrammel replied:
Use backticks for code elements

smowton reviewed (approved):
Looks good, some suggestions

peterschrammel commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
 \ingroup module_hidden 
-\page cprover-architecture-overview CProver Architecture Overview
+\ingroup module_hidden 
```
I'd split this into separate .md files. If you give them appropriate \page directives they will appear as pages within the `CProver Developer Documentation` section.

owen-mc-diffblue replied:
Perhaps when this file starts getting too big I will split of large parts into their own pages and just link to them from this page. I like having the headings in this order, as I think this is roughly the order that a new developer would want to encounter the documentation.

peterschrammel commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+FIXME
+
+
+\section section-folder-walkthrough Folder walkthrough
```
I'd appreciate if this was kept on top of one of the pages within `CProver Developer Documentation` as it is the most convenient entry point to access the module/directory documentation.

owen-mc-diffblue replied:
I've moved it to a separate file.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 1bc437dfd5931a46b6e867da80f4282bf8832d28).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80002298

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+
+#### Martin ####
+
+FIXME
```
Do I need to be fixed? :-(

owen-mc-diffblue replied:
Ha. Yes, I think I failed to parse that properly. The outline came from Cesar and he presumably meant that you could fill in that section. I'll make that clearer.

cesaro replied:
Probably what I meant was that Martin could help with the goto-analyzer section, or that I would ask him to complete the TOC of that section

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+
+FIXME
 
-\author Martin Brain, Peter Schrammel
```
I was told not to remove authors.

owen-mc-diffblue replied:
Oops. It wasn't my intention to remove authors. I deleted one file and moved most of its contents into a new file. Git has decided to represent this as moving the old file and adding a lot of things. I have reinstated them in the new file.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -1,11 +1,196 @@
+CPROVER is written in a fairly minimalist subset of C++; templates and
+meta-programming are avoided except where necessary.
+External library dependencies are avoided (only STL and a SAT solver
+are required). Boost is not used. The `util` directory contains many
```
Where does this text come from?  It reminds me of some of my usual grammatical mistakes.

owen-mc-diffblue replied:
Yes, I believe it is by you. It was in `cprover-archicture-overview.md`, which git has decided was renamed to `CBMC-developer-guide.md` in this PR. I think I moved it into a different section to match the order I was planning on.

martin-cs reviewed (commented):
<pessimistic>
Documentation marked "FIXME" won't get fixed.  We either need a complete patch or we need someone to "own" and have responsibility for documentation. 
</pessimistic>

Forgive me rambling but here is some context.  I pushed Daniel to start the group wiki and got people to write some stuff for it (mostly still relevant but it was abandoned in the move to Diffblue).  i also wrote the original internals documentation in a deal with @tautschnig .  I got part way through writing the update but left it with @Charliemowood when I was reassigned.  Some of it seems to have been committed by @peterschrammel but quite a bit appears to have been cut.

If this is the start of someone taking ownership of the documentation -- great.  If it is an optimisitic "people will help out" ... it doesn't happen organically, or, at least, I never succeeded making it happen organically.  I'd even be happy to help with the documentation push but I don't have time to lead a campaign.

peterschrammel commented:
> but quite a bit appears to have been cut.

@martin-cs, with Charlie I managed to get all the various sources of documentation integrated into http://cprover.diffblue.com. But we didn't get far in improving the structure and consolidating the texts back then. If you think that something is missing then please let me know.

owen-mc-diffblue commented:
@martin-cs Thanks for your review. There is quite a lot of energy behind improving the documentation. The idea behind creating all the headers without most of the content is to say "This is the structure the documentation should take" so that people can work together to fill it in. I only really put the "FIXME" lines in to separate the headings and make it clearer what they were meant to be. I doubt that they will all be filled in in the next few months, but I believe that some of them will be. I've replaced them with "To be documented.", which I think better portrays my intent.

Commit:
Address review comments

owen-mc-diffblue commented:
@martin-cs I forgot to say that I have 1 day a week assigned to improving documentation and onboarding material for Diffblue, so I do intend to fill in some of the sections myself. (Or move documentation from other places to this document.)

Commit:
Move folder walkthrough to a separate file

The CBMC Developer Guide now contains a link to that file.

owen-mc-diffblue commented:
I believe I've addressed all review comments and this is ready for re-review.

Commit:
Fix links between files

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8d5ac3b83f77928da252e5df9dbb71f789a51b76).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80117571
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a88f851904001fec3ab0d97ffdca56aca9171a89).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80118390

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 70f0ad6dca2770237d8368e356087fedf8a5e1a4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80121705

owen-mc-diffblue commented:
@peterschrammel @martin-cs Please re-review.

peterschrammel commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+
+### goto-analyzer ###
+
+To be documented. (Martin?)
```
Remove `(Martin?)`

martin-cs replied:
I don't mind being volunteered for things.

peterschrammel commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+
+### cbmc ###
+
+See <a href="https://github.com/diffblue/doc-diffblue-common/wiki/Talks#cbmc-for-coverage-test-generation-a-brief-tour">
```
This link is useless because not publicly accessible.

peterschrammel commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+`union_typet`.
+
+
+\section section-background-concepts Background Concepts
```
I'd put this section into a separate file

peterschrammel commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+
+
+
+\section section-other-tools Other Tools
```
Also, this section should probably become a separate file.

Commit:
Move other tools into a separate file

martin-cs commented:
On Thu, 2018-07-26 at 06:55 -0700, owen-jones-diffblue wrote:
> @martin-cs Thanks for your review. There is quite a lot of energy
> behind improving the documentation.

Great!  It's really important but so rarely are people willing to
commit time.

>  The idea behind creating all the headers without most of the content
> is to say "This is the structure the documentation should take" so
> that people can work together to fill it in.

From my experience, the best way of achieving this is to target
particular people and approach them directly.

>  I only really put the "FIXME" lines in to separate the headings and
> make it clearer what they were meant to be. I doubt that they will
> all be filled in in the next few months, but I believe that some of
> them will be. I've replaced them with "To be documented.", which I
> think better portrays my intent.

Fair.  Target specific people.  Trading time is always a good tactic,
as in "what can I do for an hour to give you time to write this
section".

> @martin-cs I forgot to say that I have 1 day a week assigned to
> improving documentation and onboarding material for Diffblue, so I do
> intend to fill in some of the sections myself. (Or move documentation
> from other places to this document.)

Brilliant!  I'll dig out some notes to send to you.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+
+Apart from the (user-orientated) CBMC user manual and this document, most
+of the rest of the documentation is inline in the code as `doxygen` and
+some comments. A man page for CBMC, goto-cc and goto-instrument is
```
At one point we even used to have a link to the live doxygen on the web.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+some comments. A man page for CBMC, goto-cc and goto-instrument is
+contained in the `doc/` directory and gives some options for these
+tools. All of these could be improved and patches are very welcome. In
+some cases the algorithms used are described in the relevant papers.
```
Last few lines aren't about doxygen and should be in a separate section.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+interpreted) are named with `_typet`. For example `ui_message_handlert`
+rather than `UI_message_handlert` or `UIMessageHandler` and
+`union_typet`.
+
```
I'm unsure whether this should be here or in the coding style document itself.  My feeling is that the coding style documents and the lint script that enforces them should be aligned.

owen-mc-diffblue replied:
It should probably be in the coding standard. I'm going to leave it for now, to try and get this PR merged more quickly.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+## Static analysis ##
+
+To be documented.
+
```
This is quite a mix of representations and analysis techniques.  I'd separate the two.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+
+\section section-cprover-architecture CPROVER Architecture
+
+This section provides a graphical overview of how CBMC fits together.
```
CBMC \subseteq CPROVER  This is the architecture of CBMC, not the architecture of CPROVER.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+
+\section section-folder-walkthrough Folder walkthrough
+
+See [the folder walkthrough](\ref folder-walkthrough).
```
When explaining this, I try to work from the architecture diagram and then show how the architectural components map on to the directories.

owen-mc-diffblue replied:
Yes, it isn't ideal having them on different pages. I'm going to leave it like this for now but it may well be changed in the future as more documentation is written.

martin-cs commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+To be documented.
+
+
+## Examples: how to represent the AST of statements, focus on java ##
```
Why focus on Java?

owen-mc-diffblue replied:
I've changed it to "how to represent the AST of statements, in C and in java"

martin-cs reviewed (approved):
Editing comments mostly.

cesaro commented on doc/architectural/CBMC-developer-guide.md:
```diff
@@ -0,0 +1,507 @@
+To be documented.
+
+## Generating doxygen documentation ##
+
```
If the latest version of this is going to be available in http://cprover.diffblue.com/ it should be worth mentioning it here.

Commit:
Rearrange everything into separate pages

Commit:
Address review comments

owen-mc-diffblue commented:
I believe I've addressed most review comments. The ones I didn't address had merit but I will leave them for the future, in the hope of getting this PR merged sooner.

@peterschrammel I have rearranged things so the major sections of the CBMC developer guide now appear as pages under "CProver Developer Documentation" (in the correct order), and there is no longer a page for the CBMC developer guide.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 612b4f8887d902c00102b223ab0fb6eef8cb1f4f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80903005

owen-mc-diffblue commented:
@martin-cs Yes, if you have notes that you want turned into documentation then please send them to me and I'll give it a go.

martin-cs commented:
@owen-jones-diffblue : sent by e-mail

owen-mc-diffblue commented:
@peterschrammel Would you like to look at this again or can I merge it?

Charliemowood commented:
Nice to see this done. :)

Edit: or shall we say living.
